### PR TITLE
[Inference/SpecDec] Add Speculative Decoding Implementation

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

Part of #5245 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Add implementation for speculative-decoding steps with accommodations on triton kernels as well as modeling.
 
Benchmarking with GSM8K dataset, for batch size 1:
Total number of sequences: 195
Elapsed time: 111.7032 s
Total out tokens: 12176
Throughput: 109.0032 tokens/s

---

image

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/batch_bucket.py [hunk 1]
```diff
@@ -42,6 +42,9 @@ def __init__(
         self.device = device or get_current_device()
         self.dtype = dtype
 
+        self._use_spec_dec = False
+        self._ver_num = None
```
Comment (FrankLeeeee) [inline colossalai/inference/batch_bucket.py:46]: What is `ver_num`? version number?

Comment (yuanheng-zhao) [reply colossalai/inference/batch_bucket.py:46]: Verification number, or num_to_verify: it's the number of tokens generated by the small/drafter model to verify in the context of large/main model as a verifier. Might revise to name to something like num_to_verify?

File: colossalai/inference/batch_bucket.py [hunk 2]
```diff
@@ -88,6 +91,23 @@ def is_compact(self):
...
+        return self._use_spec_dec
+
+    @property
+    def ver_num(self) -> int:
+        assert self.use_spec_dec and self._ver_num is not None
+        return self._ver_num
+
+    def set_use_spec_dec(self, n: int = 5) -> None:
+        self._use_spec_dec = True
+        self._ver_num = n
```
Comment (FrankLeeeee) [inline colossalai/inference/batch_bucket.py:105]: give the argument n the same name as self._ver_num

File: colossalai/inference/batch_bucket.py (hunk 1 above)
Comment (yuanheng-zhao) [reply colossalai/inference/batch_bucket.py]: Renamed.

File: colossalai/inference/batch_bucket.py (hunk 2 above)
Comment (yuanheng-zhao) [reply colossalai/inference/batch_bucket.py]: Renamed

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**